File: device/battery/android/java/src/org/chromium/device/battery/BatteryMonitorImpl.java
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.device.battery;

import org.chromium.mojo.system.MojoException;
import org.chromium.mojom.device.BatteryMonitor;
import org.chromium.mojom.device.BatteryStatus;

import java.util.ArrayList;
import java.util.List;

/**
 * Android implementation of the battery monitor service defined in
 * device/battery/battery_monitor.mojom.
 */
public class BatteryMonitorImpl implements BatteryMonitor {
    // Factory that created this instance and notifies it about battery status changes.
    private final BatteryMonitorFactory mFactory;
    private final List<QueryNextStatusResponse> mCallbacks;
    private BatteryStatus mStatus;
    private boolean mHasStatusToReport;
    private boolean mSubscribed;

    public BatteryMonitorImpl(BatteryMonitorFactory batteryMonitorFactory) {
        mFactory = batteryMonitorFactory;
        mCallbacks = new ArrayList<QueryNextStatusResponse>();
        mHasStatusToReport = false;
        mSubscribed = true;
    }

    private void unsubscribe() {
        if (mSubscribed) {
            mFactory.unsubscribe(this);
            mSubscribed = false;
        }
    }

    @Override
    public void close() {
        unsubscribe();
    }

    @Override
    public void onConnectionError(MojoException e) {
        unsubscribe();
    }

    @Override
    public void queryNextStatus(QueryNextStatusResponse callback) {
        mCallbacks.add(callback);

        if (mHasStatusToReport) {
            reportStatus();
        }
    }

    void didChange(BatteryStatus batteryStatus) {
        mStatus = batteryStatus;
        mHasStatusToReport = true;

        if (!mCallbacks.isEmpty()) {
            reportStatus();
        }
    }

    void reportStatus() {
        for (QueryNextStatusResponse callback : mCallbacks) {
            callback.call(mStatus);
        }
        mCallbacks.clear();
        mHasStatusToReport = false;
    }
}

File: services/device/battery/battery_monitor_impl.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "services/device/battery/battery_monitor_impl.h"

#include <utility>

#include "base/bind.h"
#include "base/logging.h"
#include "base/memory/ptr_util.h"
#include "mojo/public/cpp/bindings/strong_binding.h"

namespace device {

// static
void BatteryMonitorImpl::Create(mojom::BatteryMonitorRequest request) {
  auto* impl = new BatteryMonitorImpl;
  auto binding =
      mojo::MakeStrongBinding(base::WrapUnique(impl), std::move(request));
  impl->binding_ = binding;
}

BatteryMonitorImpl::BatteryMonitorImpl() : status_to_report_(false) {
  // NOTE: DidChange may be called before AddCallback returns. This is done to
  // report current status.
  subscription_ = BatteryStatusService::GetInstance()->AddCallback(
      base::Bind(&BatteryMonitorImpl::DidChange, base::Unretained(this)));
}

BatteryMonitorImpl::~BatteryMonitorImpl() {}

void BatteryMonitorImpl::QueryNextStatus(
    const QueryNextStatusCallback& callback) {
  if (!callback_.is_null()) {
    DVLOG(1) << "Overlapped call to QueryNextStatus!";
    binding_->Close();
    return;
  }
  callback_ = callback;

  if (status_to_report_)
    ReportStatus();
}

void BatteryMonitorImpl::RegisterSubscription() {}

void BatteryMonitorImpl::DidChange(const mojom::BatteryStatus& battery_status) {
  status_ = battery_status;
  status_to_report_ = true;

  if (!callback_.is_null())
    ReportStatus();
}

void BatteryMonitorImpl::ReportStatus() {
  callback_.Run(status_.Clone());
  callback_.Reset();

  status_to_report_ = false;
}

}  // namespace device
